### PR TITLE
bpf,config: Add ENABLE_LOCAL_REDIRECT_POLICY macro

### PR DESCRIPTION
**Pull request by aditighag**

### CODE REVIEW

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -235,6 +235,7 @@ sock4_wildcard_lookup_full(struct lb4_key *key __maybe_unused,
 	return NULL;
 }
 
+#ifdef ENABLE_LOCAL_REDIRECT_POLICY
```
Comment (dylandreimerink) [inline bpf/bpf_sock.c:238]: I believe you don't need this `#ifdef`, its always inlined and you already have an `#ifdef` at the call site. We want to 
keep the amount of `#ifdef`s in the code down if we can so going clang-free is easier.

File: bpf/lib/lb.h [hunk 2]
```diff
@@ -342,11 +342,13 @@ bool lb6_svc_is_routable(const struct lb6_service *svc)
 	return __lb_svc_is_routable(svc->flags);
 }
 
+#ifdef ENABLE_LOCAL_REDIRECT_POLICY
```
Comment (dylandreimerink) [inline bpf/lib/lb.h:345]: Same, here, I don't think we strictly need this `#ifdef`

**Review by dylandreimerink — CHANGES_REQUESTED** [review]

File: bpf/bpf_sock.c (hunk 1 above)
Comment (aditighag) [reply bpf/bpf_sock.c:238]: As I thought, clang complains about unused code if I removed those ifdefs - 

```
bpf_sock.c:258:1: error: unused function 'sock4_skip_xlate_if_same_netns' [-Werror,-Wunused-function]
sock4_skip_xlate_if_same_netns(struct bpf_sock_addr *ctx __maybe_unused,
```

Comment (aditighag) [reply bpf/bpf_sock.c:238]: Reverted to the original changes because of failures.

Comment (dylandreimerink) [reply bpf/bpf_sock.c:238]: Ah, of-course

**Review by dylandreimerink — APPROVED** [review]

**Comment by aditighag** [comment]
ci-clustermesh hit - https://github.com/cilium/cilium/issues/29856. 
failure in ci-ginkgo - LVH VM provisioning

None of the test failures are related to code changes so marking it as ready for merge.

**Comment by joestringer** [comment]
Please make CI green.

If we're hitting recurring problems with CI that makes it fail consistently then please help raise awareness and drive improvements to CI. Committers shouldn't be assigning `ready-to-merge` to bypass the CI, this is how we make CI worse for everyone over time because everyone just ignores CI. If it's so unreliable that we're not even able to do a quick re-run of the tests to make them pass, then we have a more fundamental issue that won't be solved by ignoring the problem and getting tophat to bypass it.

**Comment by aditighag** [comment]
@joestringer Please note that it wasn't required to run the full CI on this PR as the changes were very specific to a feature. I did highlight the failures in my last comment in case you missed it - https://github.com/cilium/cilium/pull/31098#issuecomment-1977378043. 

It's futile to re-rerun those tests, so I'm not sure what the expectation here is?

**Comment by joestringer** [comment]
I see what you mean. The documented steps under [Getting a Pull Request Merged](
https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#getting-a-pull-request-merged) -> [CI Failure Triage](https://docs.cilium.io/en/stable/contributing/testing/ci/#ci-failure-triage) pretty much describe exactly what you said. 

Recently though we have been improving CI (for instance to avoid running unnecessary jobs) and I've seen committers discuss a "re-run until green" (or fix it if it's broken) policy. The core idea is that when we ignore CI, we train ourselves that it does not provide valuable signal, which makes us (a) more likely to ignore it even when it provides legitimate signal, and (b) less likely to do something about the state of CI when it's broken. The consequences of (a) that we have experienced several times last release was that people ignore CI and then break the main branch for *everyone* for a day or two. Over time (b) leads to everyone thinking that CI is "someone else's problem" even though every contributor suffers. 

From my perspective, the rules would be very simple:
- All codeowners must approve
- CI is passing for all required jobs is passing

Furthermore, as I understand you're a committer and therefore should have access to click the "Merge when ready" button at the bottom of the PR, _as long as_ CI passes and reviews are in. This means that if you are willing to babysit the CI on your PR, you can get your PRs in without any coordination with the @cilium/tophat . This frees up the @cilium/tophat to help other community members.

That said, I think we can improve the docs and ensure that the contributor community is on-board with this policy to ensure that we're applying these rules consistently across PRs in the community.

**Comment by joestringer** [comment]
I also see a third failure under [E2E Test (1.29, f14-datapath-service-ns-xdp-2)](https://github.com/cilium/cilium/actions/runs/8145517261/job/22261994948#logs) which also doesn't look related to your PR, but would be worth triaging to at least see whether it's been reported already.

**Comment by aditighag** [comment]
>I also see a third failure under [E2E Test (1.29, f14-datapath-service-ns-xdp-2)](https://github.com/cilium/cilium/actions/runs/8145517261/job/22261994948#logs) which also doesn't look related to your PR, but would be worth triaging to at least see whether it's been reported already.

Ah, I missed that one. Filed - https://github.com/cilium/cilium/issues/31152

**Pull request merged**